go/saltpack/sign_stream.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package saltpack
+
+import (
+	"bytes"
+	"io"
+)
+
+type signAttachedStream struct {
+	header      *SignatureHeader
+	wroteHeader bool
+	encoder     encoder
+	buffer      bytes.Buffer
+	block       []byte
+	seqno       PacketSeqno
+	secretKey   SigningSecretKey
+}
+
+func newSignAttachedStream(w io.Writer, signer SigningSecretKey) (*signAttachedStream, error) {
+	if signer == nil {
+		return nil, ErrInvalidParameter{message: "no signing key provided"}
+	}
+
+	header, err := newSignatureHeader(signer.PublicKey(), MessageTypeAttachedSignature)
+	if err != nil {
+		return nil, err
+	}
+
+	stream := &signAttachedStream{
+		header:    header,
+		encoder:   newEncoder(w),
+		block:     make([]byte, SignatureBlockSize),
+		secretKey: signer,
+	}
+
+	return stream, nil
+}
+
+func (s *signAttachedStream) Write(p []byte) (int, error) {
+	if !s.wroteHeader {
+		s.wroteHeader = true
```

Comment:
Maybe set this flag after we've checked the error? It would be weird if the caller ignored the error, and then subsequent calls to Write skipped this block and started succeeding.